File: backend/app/schemas/incident_report.py
"""Pydantic schemas for incident reports."""

from datetime import date, datetime, time

from pydantic import BaseModel, ConfigDict

from app.schemas.dispatch_call import Coordinates


class IncidentReportOut(BaseModel):
    """Incident report response schema."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    incident_id: str
    incident_number: str | None = None

    incident_category: str | None = None
    incident_subcategory: str | None = None
    incident_description: str | None = None
    resolution: str | None = None

    incident_date: date | None = None
    incident_time: time | None = None
    report_datetime: datetime | None = None

    coordinates: Coordinates | None = None
    location_text: str | None = None
    police_district: str | None = None
    analysis_neighborhood: str | None = None


class IncidentReportsResponse(BaseModel):
    """Paginated response for incident reports."""

    incidents: list[IncidentReportOut]
    next_cursor: str | None = None
    total: int | None = None


class IncidentSearchParams(BaseModel):
    """Parameters for searching incident reports."""

    cursor: str | None = None
    limit: int = 50
    since: datetime | None = None
    until: datetime | None = None
    district: str | None = None
    category: str | None = None

File: ios/SFCrime/Domain/Models/IncidentReport.swift
import Foundation
import CoreLocation

/// Represents a historical police incident report from DataSF.
/// These are filed reports that appear 24-72 hours after the incident.
struct IncidentReport: Identifiable, Codable, Hashable {
    let id: Int
    let incidentId: String
    let incidentNumber: String?

    let incidentCategory: String?
    let incidentSubcategory: String?
    let incidentDescription: String?
    let resolution: String?

    let incidentDate: Date?
    let incidentTime: String?  // Time as string since API returns "HH:mm"
    let reportDatetime: Date?

    let coordinates: Coordinates?
    let locationText: String?
    let policeDistrict: String?
    let analysisNeighborhood: String?

    /// Geographic coordinates for map display
    struct Coordinates: Codable, Hashable {
        let latitude: Double
        let longitude: Double

        var clLocationCoordinate: CLLocationCoordinate2D {
            CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        }
    }

    enum CodingKeys: String, CodingKey {
        case id
        case incidentId = "incident_id"
        case incidentNumber = "incident_number"
        case incidentCategory = "incident_category"
        case incidentSubcategory = "incident_subcategory"
        case incidentDescription = "incident_description"
        case resolution
        case incidentDate = "incident_date"
        case incidentTime = "incident_time"
        case reportDatetime = "report_datetime"
        case coordinates
        case locationText = "location_text"
        case policeDistrict = "police_district"
        case analysisNeighborhood = "analysis_neighborhood"
    }

    /// Formatted date string for display
    var formattedDate: String {
        guard let date = incidentDate else { return "Unknown date" }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        return formatter.string(from: date)
    }

    /// Category display with fallback
    var displayCategory: String {
        incidentCategory ?? "Unknown"
    }
}

// MARK: - API Response

/// Paginated response for incident reports
struct IncidentReportsResponse: Codable {
    let incidents: [IncidentReport]
    let nextCursor: String?
    let total: Int?

    enum CodingKeys: String, CodingKey {
        case incidents
        case nextCursor = "next_cursor"
        case total
    }
}
